### PR TITLE
authz support string contains match

### PR DESCRIPTION
**Pull request by wulianglongrd**
support string contains match or authz `request.headers`, fix https://github.com/istio/istio/issues/52557

This is probably a behavior change: 
the old unit tests treat it as a `suffix` match, this PR treat it as a `contains` match, see:
https://github.com/istio/istio/blob/42e7e83686df24d519c5dd6ac626e3394e44c61b/pilot/pkg/security/authz/matcher/header_test.go#L48

Another option is to treat `.*foo.*` (rather than `*foo*`) as a `contains` semantic. (Or use other methods to support `contains` or `regex` semantics.)

### CODE REVIEW

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 1]
```diff
@@ -45,11 +45,11 @@ func TestHeaderMatcher(t *testing.T) {
 		{
 			Name: "suffix match",
 			K:    ":path",
-			V:    "*/productpage*",
+			V:    "*/productpage",
```
Comment (wulianglongrd) [inline pilot/pkg/security/authz/matcher/header_test.go:48]: the old unit tests treat it as a `suffix` match, this PR treat it as a `contains` match(see L81).

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
I think the right solution should to support regex match

**Comment by wulianglongrd** [comment]
> I think the right solution should to support regex match

Thanks for the review, I have similar ideas to yours, but I currently don't have a good way to express regex semantics. 

Do you have any suggestions for using [key or values](https://istio.io/latest/docs/reference/config/security/authorization-policy/#Condition) to express regex semantics?

**Comment by wulianglongrd** [comment]
BTW, currently values support the following semantics:
1. `*`: preset match
2. `*{anything}`: suffix match
3. `{anything}*`: prefix match
4. the rest: exact match

Use `regex://` prefix to express regex semantics?

**Comment by hzxuzhonghu** [comment]
seems too hardcoding, how about adding a new field type []StringMatcher like what we do in VS, this field to deprecate the old `values`

**Comment by howardjohn** [comment]
Thanks for the PR. I have a few concerns

1. As this is an API change to a v1 API, this should go through API review before implementation 
2. This seems inconsistent to apply only to headers
3. The intended use case here seems to be matching multiple headers flattened into a single one, but doesn't actually do that securely - I worry this would lead to easily miswritten policies that introduce vulnerability

**Comment by wulianglongrd** [comment]
> seems too hardcoding, how about adding a new field type []StringMatcher like what we do in VS, this field to deprecate the old `values`

Thanks, this solution makes sense to me.

**Comment by wulianglongrd** [comment]
Thanks.

> 1. As this is an API change to a v1 API, this should go through API review before implementation

agree.

> 2. This seems inconsistent to apply only to headers

Thanks for pointing this out, it also applies to `:method` and probably nothing else.

> 3. The intended use case here seems to be matching multiple headers flattened into a single one, but doesn't actually do that securely - I worry this would lead to easily miswritten policies that introduce vulnerability

 😊 Can you give an example for this?

**Comment by howardjohn** [comment]
> 😊 Can you give an example for this?

Yes, sorry I sent that from my phone so was hard to elaborate..

In your example:
* rule is `*foo=*`. `"Cookie: not-foo=bar"` unexpecedly matches
* rule is `*foo*`. `"X-foo: not foo bar"` unexpectedly matches

**Comment by wulianglongrd** [comment]
@hzxuzhonghu @howardjohn 

How about using the following syntax? One of `values`, `notValues`, `match`. The new `match` field is `StringMatch` type.

```yaml
  rules:
  - when:
    # old syntax
    - key: request.headers[header]
      values: ["value"]

    # new syntax: exact match
    - key: request.headers[header1]
      match:
        exact: value1

    # new syntax: prefix match
    - key: request.headers[header2]
      match:
        prefix: value2

    # new syntax: regex match
    - key: request.headers[header3]
      match:
        regex: value3
```

**Comment by hzxuzhonghu** [comment]
>Note: at least one of `values`, `notValues` or `match` must be set.

only one or at least one?

**Comment by wulianglongrd** [comment]
> > Note: at least one of `values`, `notValues` or `match` must be set.
> 
> only one or at least one?

a. values, not_values
b. match

At least one of `a` or `b`. when `a` is selected, at least one of `values` or `not_values`.

**Comment by howardjohn** [comment]
In that past we have had pretty conclusive decision to not add regex into authz policy. See discussion in https://github.com/istio/istio/issues/25021 and many linked issues.

Even to serve this use case, regex is inappropriate as I tried to show in https://github.com/istio/istio/pull/52583#issuecomment-2284453746.

**Comment by howardjohn** [comment]
What is really desired here, it seems, is to be able to match on individual headers instead of merged ones

**Comment by hzxuzhonghu** [comment]
>>Note: at least one of values, notValues or match must be set.

>only one or at least one?

@wulianglongrd I think it should be only one, not more than one. Like you cannot specify string match with values, they can conflict

**Comment by hzxuzhonghu** [comment]
>What is really desired here, it seems, is to be able to match on individual headers instead of merged ones

@howardjohn regex can be used to match this with `;` provided, for more complex case, it may need envoy to support

**Comment by wulianglongrd** [comment]
Sorry for the late reply.

> In that past we have had pretty conclusive decision to not add regex into authz policy. See discussion in #25021 and many linked issues.

Thanks @howardjohn . I take a quick look and saw some reasons for not implementing regex: https://github.com/istio/istio/issues/16585#issuecomment-912846421, https://github.com/istio/istio/issues/16585#issuecomment-1456359549, https://github.com/istio/istio/issues/16585#issuecomment-1164633757, https://github.com/istio/istio/issues/16585#issuecomment-984790462 . BTW, it seems that the `path templating matching`(https://github.com/istio/istio/pull/50365) currently only supports `path`, which does not cover https://github.com/istio/istio/issues/52557 issue (`request.headers`).


> Even to serve this use case, regex is inappropriate as I tried to show in #52583 (comment).

I think correct regex can prevent these issues. But, inaccurate regex may cause security issues.


> What is really desired here

The motivation is to fix https://github.com/istio/istio/issues/52557 and hope that the matching rules of authz and VS are consistent.


> I think it should be only one, not more than one. Like you cannot specify string match with values, they can conflict

Thanks @hzxuzhonghu . Yes, if it is "more than one", the usage will be more complicated and there may be conflicts.

**Comment by howardjohn** [comment]
#52557 is really a structured match requirement.. If we had a way to match on individual split headers instead of aggregated ones, it would solve this.

That is, instead of matching against `cookie1=whatever; cookie2=whatever;`, it would match against `cookie1=whatever` and `cookie2=whatever`. Not sure what this would look like in terms of API, AND vs OR them together, etc though.

**Pull request closed without merging**